blueprints/user.permissions.md

```diff
@@ -0,0 +1,286 @@
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+# Enforced User Permissions
+
+## Problem Description
+Currently, the Permissions (currently "Capabilities", henceforth referred to as
+"Permissions" to avoid confusion with Server Capabilities) afforded to a user
+are defined by the "Privilege Level" of their Role. It essentially defines a
+scale on the interval [0,30], where a higher number allows more things. This
+isn't as scalable or configurable as it could be, and involves granting more
+permissions to a user than they actually need just because every Privilege
+Level encapsulates all of the permissions of every Privilege Level below it,
+making disjoint Permission sets for disparate Roles logically impossible.
+
+## Proposed Change
+Instead of a system of rigid supersets of Permissions, this blueprint proposes
+that the system of user "Capabilities" be expanded upon, renamed to Permissions
+(again, to avoid confusion) and enforced by the API.
+
+Under this system, each endpoint will declare the Permissions it requires for
+specific operations and check that the requesting user has those required, if
+any.
+
+This information is _not_ to be stored in the database nor exposed through the
+API, but documented. This change is discussed in greater detail in the
+"Alternatives" section.
+
+The enforcement of these Permissions is proposed to be configurable via an
+entry in the Traffic Ops configuration file. Thus, for one major version the
+API will need to support Permissions-based authorization as well as Privilege
+Level-based authorization.
+
+## Data Model Impact
+"Capabilities" as they are known today will be removed from the data model.
+Also, the current model of a Role will need to be modified. The current model
+is as follows:
+
+```typescript
+interface Role {
+	capabilities: Array<string> | null;
+	description: string | null;
+	id?: number | null;
+	name: string | null;
+	privLevel: number | null;
+}
+```
+
+The proposed new model is shown below.
+
+```typescript
+interface Role {
+	description: string;
+	lastUpdated?: Date;
+	name: string;
```

Comment:
I thought someone said that roles will be per-tenant. Doesn't this mean it should have a `tenantId`?

Reply:
at one time, i did think roles could/should be scoped per tenant but i'm kinda thinking that is overkill atm. but it is something to think about. without tenancy, any new role that is created will be visible to all "tenants". is that ok? maybe if creating new roles is a semi-rare event...

Reply:
anyhow, easier to implement now than later if that seems important.

Reply:
I think the issue there was because users in Tenant A can modify the permissions, effectively, of users in Tenant B even though users in Tenant A are not, in general, able to access Tenant B.

We could make Roles "Tenantable", but IMO that's somewhat separate to this notion of using Permissions instead of privilege level. Plus, you wouldn't need to make Roles "Tenantable" to prevent that, although it could possibly be more annoying without it.

Reply:
I don't think Roles should be tenantable; I just thought that's what someone said the current design was. But, it looks like the design has changed for the better. Disregard :)